File: wrapper_and_examples/SLICKER.m
function [time, central_tendency, uncertainty, spread] = SLICKER(target,rtime,proxy1,varargin)

% SLICKER - runs Fortran SLICKER reconstruction code "reconstruct"
%
%   Brief description : This is a Matlab wrapper that generates input files to run the Fortran
%                       SLICKER reconstruction code described in Roberts et al. (2023).
%
%   User inputs:
%       target     2d array; 1st column is time, second column is the corresponding data value
%       rtime      1d vector of list of times that you want the reconstruction for
%       proxy1     2d array; 1st column is time, 2nd column is proxy value
%       varargin   up to an additional 9 input proxy time series (proxy2, ..., proxy9) and a list of name-value pairs (listed below) to change defaults
%
%   Input arrays may be different lengths. Data for each MUST be in strictly increasing time order.
%
%   Name-value pairs and default values
%       num_ensemble    Number of ensemble members (default 4096)
%       slick_width     Slick width parameters (default [0.4 1.6])
%       time_limit      Time limit (seconds) for solution of individual 
%                        ensemble member (default 60 seconds)
%       tol             Solution tolerance (default 1e-6)
%       nonlinear       Attempt a non-linear solution (default "n")
%       frac_subset     Fraction of ensemble with best stationarity used in
%                        reported solution (default 0.5)
%       proxy_inversion Test for significantly stronger correlation if 
%                        individual proxies are inverted (default "n" for 
%                        linear solution and "y" for non-linear)
%
%
%   Usage:
%       [time, central_tendency, uncertainty, spread] = SLICKER(target,rtime,proxy1,varargin)
%
%
%   Examples:
%       [time, central_tendency, uncertainty, spread] = SLICKER(target,rtime,proxy1,'num_ensemble',512);
%       [time, central_tendency, uncertainty, spread] = SLICKER(target,rtime,proxy1,proxy2,proxy3,'time_limit',20,'tol',1e-7);
%
%
%   Citations:
%   If you use this Matlab wrapper, please cite
%   Roberts et al. (submitted), 2023.
%
%   Author Info
%   This function was written by Felicity S. McCormack of Monash University,
%   and Jason Roberts of the Australian Antarctic Division
%   Version 2.0: May 2023
%

% Some checks
p=inputParser;
validScalarPosNum = @(x) isnumeric(x) && isscalar(x) && (x > 0);
validMatrix = @(x) isnumeric(x) && ismatrix(x);
validVector = @(x) isnumeric(x) && isvector(x);
validVectorPosNum = @(x) isnumeric(x) && isvector(x) && (x > 0);
validyn = @(x) ischar(x) && (x=="y" || x=="Y" || x=="n" || x=="N");
addRequired(p,'target',validMatrix);
addRequired(p,'rtime',validVector);
addRequired(p,'proxy1',validMatrix);
addOptional(p,'proxy2',[],validMatrix);
addOptional(p,'proxy3',[],validMatrix);
addOptional(p,'proxy4',[],validMatrix);
addOptional(p,'proxy5',[],validMatrix);
addOptional(p,'proxy6',[],validMatrix);
addOptional(p,'proxy7',[],validMatrix);
addOptional(p,'proxy8',[],validMatrix);
addOptional(p,'proxy9',[],validMatrix);
addOptional(p,'proxy10',[],validMatrix);
addParameter(p,'num_ensemble',4096,validScalarPosNum);
addParameter(p,'slick_width', [0.4 1.6] ,validVectorPosNum);
addParameter(p,'time_limit', 60 , validScalarPosNum);
addParameter(p,'tol', 1e-6 , validScalarPosNum);
addParameter(p,'nonlinear', "n", validyn);
addParameter(p,'frac_subset', 0.5 , validScalarPosNum);
addParameter(p,'proxy_inversion', "e" , validyn);
parse(p,target,rtime,proxy1,varargin{:});

num_ensemble=p.Results.num_ensemble;
slick_width=p.Results.slick_width;
time_limit=p.Results.time_limit;
tol=p.Results.tol;
isnonlinear=lower(p.Results.nonlinear);
frac_subset=p.Results.frac_subset;
if (p.Results.proxy_inversion=="e")
  if (isnonlinear=="y")
    ispinversion="y";
  else
    ispinversion="n";
  end
else
  ispinversion=lower(p.Results.proxy_inversion);
end
proxy{1}=proxy1;
num_proxy_files=1;

% Get proxies 
fields = fieldnames(p.Results);
for i=4:11
	fieldName = fields{i};
	fieldValue = p.Results.(fieldName);

	if ~isempty(fieldValue)
		num_proxy_files = num_proxy_files + 1;
		proxy{num_proxy_files} = p.Results.(fieldName);
	end
end

if (nargin<3)
   help SLICKER
   error('SLICKER error message: bad usage')
end
if size(target, 2) ~= 2
    error('SLICKER error message: reconstruction target data should be Mx2.');
end

for i=1:num_proxy_files
  if size(proxy{i},2) ~= 2
    error('SLICKER error message: reconstruction proxy ',string(i),' should be Mx2.');
  end
end


% Generate user input data

% Filename of reconstruction target data
basename=strcat("SLICKER_",date,"_",datestr(now,'HH_MM_SS_FFF'),"_");
rfilename = strcat(basename,"target.txt");

% Filenames of proxy files
for i = 1:num_proxy_files
    pfilenames{i} = strcat(basename,"proxy",int2str(i),".txt");
end

% Filename of times to be used for reconstruction
tfilename = strcat(basename,"timebase.txt");


% Filename of output file
ofilename = strcat(basename,"output.txt");

% Filename of master input file
mfilename = strcat(basename,"input_keyboard_file.txt");


% Generate user input file
fid = fopen(mfilename, 'w');
fprintf(fid, '%s\n', rfilename);
fprintf(fid, '%d\n', num_proxy_files);
for i = 1:num_proxy_files
   fprintf(fid, '%s\n', pfilenames{i});
end
fprintf(fid, '%s\n', tfilename);
fprintf(fid, '%d\n', num_ensemble);
fprintf(fid, '%f %f\n', slick_width);
fprintf(fid, '%d\n', time_limit);
fprintf(fid, '%e\n', tol);
fprintf(fid, '%s\n', ofilename);
fprintf(fid, '%s\n', isnonlinear);
fprintf(fid, '%f\n', frac_subset);
fprintf(fid, '%s\n', ispinversion);
fclose(fid);


% Write target file
dlmwrite(rfilename, target, 'delimiter', '\t');

% Write reconstruction time file
dlmwrite(tfilename, rtime, 'delimiter', '\t');

% Write proxy files
for i = 1:num_proxy_files
    dlmwrite(pfilenames{i}, proxy{i}, 'delimiter', '\t');
end


% Run Fortran code
SLICK_path=which("SLICKER");
FORTRAN_path=strrep(SLICK_path,"SLICKER.m","reconstruct");
systemcommand = [FORTRAN_path,' < ', mfilename ,' > /dev/null'];
system(systemcommand);

% Read the output file 
fid = fopen(ofilename,'r');
tline=fgetl(fid);
i=1; j=1;
while ischar(tline)
   tline=fgetl(fid);
   if ischar(tline)
      index = cellfun(@(x) ~startsWith(x, '!'), {tline});
      if index
         values=strsplit(tline);
         data(j,:)=str2double({values{find(~cellfun('isempty', values))}});
         j=j+1;
      end
      i=i+1;
   end
end
fclose(fid);

time = data(:,1);
central_tendency = data(:,2);
uncertainty = data(:,3);
spread = data(:,4);

File: wrapper_and_examples/SLICKER.R
SLICKER <- function(target, rtime, p1, p2 = NULL, p3 = NULL, p4 = NULL, 
              p5 = NULL, p6 = NULL, p7 = NULL, p8 = NULL, p9 = NULL, 
              p10 = NULL, ...) {
#
#RUN_SLICKER - runs Fortran SLICKER reconstruction code "reconstruct"
#
#   Brief description : This is a R wrapper that generates input files to 
#   run the Fortran SLICKER reconstruction code described in 
#   Roberts et al. (2023).
#
#   User inputs:
#       target     2d array; 1st column is time, second column is the corresponding data value
#       rtime      1d vector of list of times that you want the reconstruction for
#       p1         2d array; 1st column is time, 2nd column is proxy value
#       p[2-10]    [optional] up to an additional 9 input proxy time series
#       ...        [optional] parameters, only needed to change defaults
#
#   Input arrays may be different lengths. Data for each MUST be in strictly increasing time order.
#
#   Optional parameters and default values
#       num_ensemble    Number of ensemble members (default 4096)
#       slick_width     Slick width parameters (default [0.4 1.6])
#       time_limit      Time limit (seconds) for solution of individual 
#                        ensemble member (default 60 seconds)
#       tol             Solution tolerance (default 1e-6)
#       nonlinear       Attempt a non-linear solution (default "n")
#       frac_subset     Fraction of ensemble with best stationarity used in
#                        reported solution (default 0.5)
#       proxy_inversion Test for significantly stronger correlation if 
#                        individual proxies are inverted (default "n" for 
#                        linear solution and "y" for non-linear)
#
#   Output:
#       a m*4 array (where m is the length of "rtime")
#           column 1 is the reconstruction times
#           column 2 is the central tendency of the SLICKER ensemble solution
#           column 3 is the 95% confidence interval of the central tendency
#           column 4 is the spread of the ensemble
#
#   Usage:
#       recon <- SLICKER(target,rtime,proxy1,proxy2)
#
#
#   Examples:
#       recon <- SLICKER(target,rtime,proxy1,num_ensemble=512)
#       recon <- SLICKER(target,rtime,proxy1,proxy2,proxy3,time_limit=20,tol=1e-7)
#
#
#   Citations:
#   If you use this R wrapper, please cite
#   Roberts et al. (submitted), 2023.
#
#   Author Info
#   This function was written by Jason Roberts and John French     
#   May 2023
#

  defaults <- list(...)

  if (!is.null(defaults$num_ensemble)) {
    num_ensemble <- defaults$num_ensemble
  } else {
    num_ensemble <- 4096
  }
  if (!is.null(defaults$slick_width)) {
    slick_width <- defaults$slick_width
  } else {
    slick_width <- c(0.4,1.6)
  }
  if (!is.null(defaults$time_limit)) {
    time_limit <- defaults$time_limit
  } else {
    time_limit <- 60
  }
  if (!is.null(defaults$tol)) {
    tol <- defaults$tol
  } else {
    tol <- 1e-6
  }
  if (!is.null(defaults$nonlinear)) {
    isnonlinear <- tolower(defaults$nonlinear)
  } else {
    isnonlinear <- "n"
  }
  if (!is.null(defaults$frac_subset)) {
    frac_subset <- defaults$frac_subset
  } else {
    frac_subset <- 0.5
  }
  if (!is.null(defaults$proxy_inversion)) {
    ispinversion <- tolower(defaults$proxy_inversion)
  } else if (isnonlinear == "y") {
    ispinversion <- "y"
  } else {
    ispinversion <- "n"
  }

  num_proxy_files <- 1
  proxy <- list()
  proxy[[1]] <- p1
  proxy_list <- list(p2, p3, p4, p5, p6, p7, p8, p9, p10)
  for (i in proxy_list) {
    if (!is.null(i)) {
      num_proxy_files <- num_proxy_files + 1
      proxy[[num_proxy_files]] <- i
    }
  }

  if (length(target) != 2) {
    stop("SLICKER error message: reconstruction target should be Mx2.")
  }

  for (i in 1:num_proxy_files) {
    if (length(proxy[[i]]) != 2) {
      stop("SLICKER error message: reconstruction proxy ", i, " should be Mx2.")
    }
  }

# Filename of reconstruction target data
basename <- paste0("SLICKER_", format(Sys.time(), "%b_%d_%Y_%H_%M_%S"), "_")
rfilename <- paste0(basename, "target.txt")

# Filenames of proxy files
cat_pfilenames <- toString(num_proxy_files)
pfilenames <- list()
for (i in 1:num_proxy_files) {
    pfilenames[[i]] <- paste0(basename, "proxy", toString(i), ".txt")
    cat_pfilenames <- paste(cat_pfilenames, pfilenames[[i]], sep = "\n")
}

# Filename of times to be used for reconstruction
tfilename <- paste0(basename, "timebase.txt")

# Filename of output file
ofilename <- paste0(basename, "output.txt")

# Filename of master input file
mfilename <- paste0(basename, "input_keyboard_file.txt")

# Generate user input file
fid <- file(mfilename)
writeLines(c(rfilename, cat_pfilenames, tfilename, toString(num_ensemble),
           toString(slick_width), toString(time_limit), toString(tol),
           ofilename, isnonlinear, toString(frac_subset), ispinversion), fid)
close(fid)

# Write target file
write.table(target, rfilename, sep = " ", row.names = FALSE, col.names = FALSE)

# Write reconstruction time file
write.table(rtime, tfilename, sep = " ", row.names = FALSE, col.names = FALSE)

# Write proxy files
for (i in 1:num_proxy_files) {
  write.table(proxy[[i]], pfilenames[[i]], sep = " ", row.names = FALSE, 
              col.names = FALSE)
}

# Run Fortran code
sys <- Sys.info()[1]

if (sys=="Windows") {
  print("OS:Windows")
  system_cmd <- paste0("..\\reconstruct_Win.exe < ", mfilename, " > console.txt")
  shell(system_cmd, wait=TRUE)
}
else if (sys=="UNIX" || sys=="Linux") {
  print("OS:UNIX")
  exec_dir <- getSrcDirectory(SLICKER)
  system_cmd <- paste0(exec_dir, "/reconstruct < ", mfilename, " > /dev/null")
  system(system_cmd)
}

# Read the output file
data <- read.delim(ofilename, header = FALSE, sep = "", comment.char = "!",
                   strip.white = TRUE)

return(data)
}


